Raise subclasses of KeyError

jamadden opened the pull request:
We've found it very useful to be able to distinguish the errors that come from an intid utility from generic dictionary/BTree errors. The main proposed errors are:

```python
class IntIdMissingError(KeyError):
        """
        Raised by the utility when ``getId`` fails.
        """

class ObjectMissingError(KeyError):
        """
        Raised by the utility when ``getObject`` fails.
        """
```

We currently handle this transformation in our own subclass, but it'd be nice if `zope.intid` did it in the first place.

I don't really see any concerns with `getId`, it was already transforming error messages.

`getObject` and `unregister` could theoretically have been raising POSKeyError before and this commit is transforming those. I'm not sure if that's a problem or not.  I could catch that error and specifically let it propagate if so (but then I'd either need to add ZODB as a dependency back, negating #3, or do a soft import.)

jamadden committed:
fix testing under py3.3 (could not find 'intid' in 'zope') and simplify

Bring in the .travis.yml changes from #3 and also do the same for tox.ini.

jamadden committed:
Raise more informative KeyError subclasses.

icemac reviewed (commented):
LGTM.
It would be nice to to propagate the `POSKeyError`. To do this you could try to import it from ZODB using a from-import. If an `ImportError` occurs create a `POSKeyError` class inside the except statement. I've seen this on some places in Zope code.

jamadden commented:
Thanks for the review!

> It would be nice to to propagate the POSKeyError. To do this you could try to import it from ZODB using a from-import. If an ImportError occurs create a POSKeyError class inside the except statement.

That's what I meant by a soft-import. I agree, you're right, that needs to propagate. I'll make the changes. The only tricky part will be testing it.

jamadden committed:
Make sure POSKeyError propagates.

Note that ``queryId`` now propagates this error; previously it was
hidden. This now matches the behaviour of ``queryObject`` which always
propagated.

Test this. I have 100% test coverage locally. Some minor refactoring of
the subscribers facilitates this (and reduces the nesting depth).

Also add a setup.cfg to indicate that we generate universal wheels.

jamadden commented:
> It would be nice to to propagate the POSKeyError.

Done.  The commit comment and CHANGES notes that now `queryId` will also propagate this (like `queryObject` has always done); previously it hid it.

Testing wasn't that hard or even that fragile, because the implementation attributes, `refs` and `ids` are both "public" and thus can't be changed.

jamadden committed:
Use latest pypy3 on travis because of pip.

icemac reviewed (approved):
LGTM.

jamadden commented:
Thanks! Can we get a PyPI release please?

jamadden commented:
@tseaver You're also a PyPI owner, could we get a release please, or if you'd like to add me (jamadden) to the PyPI owners I can do it?

tseaver commented:
@jamadden I have added `jamadden` as a PyPI owner of `zope.intid`.  "Fire when ready, Gridley!"

jamadden commented:
Awesome! Thank you.

jamadden commented:
Also, deep reference! I had to double check, but I think Gridley actually died after that order (not that it was directly at fault). So, thank you? 😉
